### PR TITLE
feat: Remove unsafe casting in favor of smart cast

### LINKED ISSUE #1832: Use kotlin smart cast and safe calls
**Actual Behaviour**

Currently, in many places, unsafe casting and Null checks(!!) are used.

**Expected Behaviour**

Use smart cast and safe calls.

**Would you like to work on the issue?**

Yes

### PR DESCRIPTION
**Pull request by liveHarshit**
Fixes #1832

### CODE REVIEW

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatActivity.kt [hunk 1]
```diff
@@ -264,10 +265,12 @@ class ChatActivity : AppCompatActivity(), IChatView {
 
     //Replies user with Speech
     override fun voiceReply(reply: String, language: String) {
-        val audioFocus = getSystemService(Context.AUDIO_SERVICE) as AudioManager
+        val mAudioFocus = getSystemService(Context.AUDIO_SERVICE)
+        var audioFocus: AudioManager? = null
+        if (mAudioFocus is AudioManager) audioFocus = mAudioFocus
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatActivity.kt]: Don't use hungarian notation. If it is not AudioManager, just return

File: app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt [hunk 2]
```diff
@@ -28,13 +28,15 @@ class STTfragment : Fragment() {
 
     override fun onAttach(context: Context?) {
         super.onAttach(context)
-        chatPresenter = ChatPresenter(activity as ChatActivity)
+        val thisActivity = activity
+        if (thisActivity is ChatActivity) chatPresenter = ChatPresenter(thisActivity)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: I think ChatPresenter requires context. Why does it require an Activity?

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: In ChatPresenter class: `class ChatPresenter(chatActivity: ChatActivity)` it is require an ChatActivity.

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatActivity.kt (hunk 1 above)
Comment (liveHarshit) [reply app/src/main/java/org/fossasia/susi/ai/chat/ChatActivity.kt]: Updated

File: app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt (hunk 2 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: What ChatActivity specific things are being used in the presenter?

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: In `private var utilModel: UtilModel = UtilModel(chatActivity)`
but UtilModel is receiving context : `class UtilModel(val context: Context) : IUtilModel`
Then it was written worng.

Should I change it in this PR after opening an issue?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: Just change it directly

Comment (liveHarshit) [reply app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: Updated

File: app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt [hunk 3]
```diff
@@ -28,13 +28,14 @@ class STTfragment : Fragment() {
 
     override fun onAttach(context: Context?) {
         super.onAttach(context)
-        chatPresenter = ChatPresenter(activity as ChatActivity)
+        if (context is Context) chatPresenter = ChatPresenter(context)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: Use `requireContext()`

**Comment by iamareebjamal** [comment]
Great work, but there is a lot of redundancy. Like we don't want to use hungarian notation and aboid duplication

**Comment by liveHarshit** [comment]
Thanks :smile: I will remove the Hungarian notations but what is it - 
> aboid duplication

**Comment by iamareebjamal** [comment]
typo

**Comment by liveHarshit** [comment]
@iamareebjamal Updated

**Comment by liveHarshit** [comment]
please suggest what else need to be changed.

File: app/src/main/java/org/fossasia/susi/ai/chat/ChatActivity.kt [hunk 4]
```diff
@@ -264,7 +265,8 @@ class ChatActivity : AppCompatActivity(), IChatView {
 
     //Replies user with Speech
     override fun voiceReply(reply: String, language: String) {
-        val audioFocus = getSystemService(Context.AUDIO_SERVICE) as AudioManager
+        val audioFocus = getSystemService(Context.AUDIO_SERVICE)
+        if (audioFocus !is AudioManager) return
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/ChatActivity.kt]: For getSystemService, it is fine to cast as exception on cast failure is accepted. After this change, error will silently fail

File: app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt [hunk 5]
```diff
@@ -28,13 +28,14 @@ class STTfragment : Fragment() {
...
-        chatPresenter = ChatPresenter(activity as ChatActivity)
+        chatPresenter = ChatPresenter(requireContext())
     }
 
     @NonNull
     override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
         val rootView = inflater.inflate(R.layout.fragment_sttframe, container, false)
-        (activity as ChatActivity).fabsetting.hide()
+        val thisActivity = activity
+        if (thisActivity is ChatActivity) thisActivity.fabsetting.hide()
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: Make this a lateinit var

File: app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/SelectionDialogListAdapter.kt [hunk 6]
```diff
@@ -20,30 +20,33 @@ import org.fossasia.susi.ai.R
...
+            if (context is Activity) view = context.layoutInflater.inflate(R.layout.item_selection_dialog_list, null)
             val viewHolder = ViewHolder()
-            viewHolder.option = view.findViewById<View>(R.id.option) as TextView
-            viewHolder.icon = view.findViewById<View>(R.id.icon) as ImageView
-            view.tag = viewHolder
+            val textView = view?.findViewById<View>(R.id.option)
+            if (textView is TextView) viewHolder.option = textView
+            val imageView = view?.findViewById<View>(R.id.icon)
+            if (imageView is ImageView) viewHolder.icon =  imageView
+            view?.tag = viewHolder
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/SelectionDialogListAdapter.kt]: This was fine as it was. Just change the generic of findViewById from View to specific

File: app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt [hunk 7]
```diff
@@ -23,7 +24,8 @@ class UtilModel(val context: Context) : IUtilModel {
     override fun saveToken(response: Response<LoginResponse>) {
         val loginResponse = response.body()
         if (loginResponse != null) {
-            PrefManager.putString(Constant.ACCESS_TOKEN, loginResponse.accessToken as String)
+            val token = loginResponse.accessToken
+            if (token is String) PrefManager.putString(Constant.ACCESS_TOKEN, token)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt]: What type is accessToken actually?

Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt]: OK, it is a nullable string. Just make PrefManager accept nullables. Then there'll be no need of this cast

File: app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt [hunk 8]
```diff
@@ -63,7 +65,7 @@ class UtilModel(val context: Context) : IUtilModel {
     }
 
     override fun setCustomURL(url: String) {
-        PrefManager.putString(Constant.CUSTOM_SERVER, CredentialHelper.getValidURL(url) as String)
+        CredentialHelper.getValidURL(url)?.let { PrefManager.putString(Constant.CUSTOM_SERVER, it) }
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/data/UtilModel.kt]: Same here

File: app/src/main/java/org/fossasia/susi/ai/data/db/DatabaseRepository.kt [hunk 9]
```diff
@@ -24,11 +24,10 @@ class DatabaseRepository : IDatabaseRepository {
...
     override fun getMessageCount(): Long {
-        val temp = realm.where(ChatMessage::class.java).max(Constant.ID)
-        return if (temp == null)
-            -1
-        else
-            temp as Long
+        val tempNumber = realm.where(ChatMessage::class.java).max(Constant.ID)
+        var tempLong: Long? = null
+        if (tempNumber is Long) tempLong = tempNumber
+        return tempLong ?: -1
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/data/db/DatabaseRepository.kt]: What is tempNumber actually? Int? Double?

Whatever it is, don't cast as Long, properly contruct it using `toLong()`

File: app/src/main/java/org/fossasia/susi/ai/device/deviceconnect/DeviceConnectFragment.kt [hunk 10]
```diff
@@ -55,8 +55,10 @@ class DeviceConnectFragment : Fragment(), IDeviceConnectView {
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
-        mainWifi = context?.getSystemService(Context.WIFI_SERVICE) as WifiManager
-        deviceConnectPresenter = DeviceConnectPresenter(activity as DeviceActivity, mainWifi)
+        val wifiService = context?.getSystemService(Context.WIFI_SERVICE)
+        if (wifiService is WifiManager) mainWifi = wifiService
+        val thisActivity = activity
+        if (thisActivity is DeviceActivity) deviceConnectPresenter = DeviceConnectPresenter(thisActivity, mainWifi)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/device/deviceconnect/DeviceConnectFragment.kt]: Why is DeviceActivity needed in presenter

File: app/src/main/java/org/fossasia/susi/ai/device/deviceconnect/DeviceConnectFragment.kt [hunk 11]
```diff
@@ -88,20 +90,22 @@ class DeviceConnectFragment : Fragment(), IDeviceConnectView {
     }
 
     override fun showLocationIntentDialog() {
-        val dialogBuilder = AlertDialog.Builder(activity as DeviceActivity)
-        dialogBuilder.setView(R.layout.select_dialog_item_material)
+        val thisActivity = activity
+        var dialogBuilder: AlertDialog.Builder? = null
+        if (thisActivity is DeviceActivity) dialogBuilder = AlertDialog.Builder(thisActivity)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/device/deviceconnect/DeviceConnectFragment.kt]: Same. You just need context

File: app/src/main/java/org/fossasia/susi/ai/helper/ChatBubbleLayout.kt [hunk 12]
```diff
@@ -86,24 +86,24 @@ class ChatBubbleLayout : FrameLayout {
...
-                (childDateView.layoutParams as FrameLayout.LayoutParams).bottomMargin = bottomMargin
+                if (layoutParams is FrameLayout.LayoutParams) layoutParams.bottomMargin = bottomMargin
             }
 
         } else {
             finalFramelayoutWidth = Math.max(finalFrameLayoutRequiredWidth,
                     childTextView.measuredWidth + viewPaddingLeftNRight)
             finalFrameLayoutHeight = measuredHeight
-            (childDateView.layoutParams as FrameLayout.LayoutParams).bottomMargin = bottomMargin
+            if (layoutParams is FrameLayout.LayoutParams) layoutParams.bottomMargin = bottomMargin
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/ChatBubbleLayout.kt]: Fine as it was. This whole file

File: app/src/main/java/org/fossasia/susi/ai/helper/DateTimeHelper.kt [hunk 13]
```diff
@@ -23,7 +23,8 @@ object DateTimeHelper {
     val currentTime: String
         get() {
             val delegate = "hh:mm aaa"
-            return DateFormat.format(delegate, Calendar.getInstance().time) as String
+            val dateTime = DateFormat.format(delegate, Calendar.getInstance().time)
+            return if (dateTime is String) dateTime else ""
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/DateTimeHelper.kt]: What is the return type of format? If it is nullable string, then use datetime ?: ""

Or nullToEmpty

File: app/src/main/java/org/fossasia/susi/ai/helper/LocationHelper.kt [hunk 14]
```diff
@@ -59,16 +59,16 @@ class LocationHelper
...
         var location: Location?
-        if (mContext == null) {
+        if (context == null) {
             return
         }
         try {
-            locationManager = mContext
-                    .getSystemService(Context.LOCATION_SERVICE) as LocationManager
+            val mLocationManager = context.getSystemService(Context.LOCATION_SERVICE)
+            if (mLocationManager is LocationManager) locationManager = mLocationManager
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/helper/LocationHelper.kt]: Remove hungarian notation, as use cast

File: app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt [hunk 15]
```diff
@@ -198,7 +198,8 @@ class LoginPresenter(loginActivity: LoginActivity) : ILoginPresenter, ILoginMode
             if (CredentialHelper.isURLValid(url)) {
                 if (CredentialHelper.getValidURL(url) != null) {
                     utilModel.setServer(false)
-                    utilModel.setCustomURL(CredentialHelper.getValidURL(url) as String)
+                    val validUrl = CredentialHelper.getValidURL(url)
+                    if (validUrl is String) utilModel.setCustomURL(validUrl)
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt]: Check for non null, not for is String

File: app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt [hunk 16]
```diff
@@ -198,7 +198,8 @@ class LoginPresenter(loginActivity: LoginActivity) : ILoginPresenter, ILoginMode
             if (CredentialHelper.isURLValid(url)) {
                 if (CredentialHelper.getValidURL(url) != null) {
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt]: This is already checked above. Refactor into a val and you won't need anything else

**Comment by iamareebjamal** [comment]
Make similar changes wherever applicable

File: app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt (hunk 5 above)
Comment (liveHarshit) [reply app/src/main/java/org/fossasia/susi/ai/chat/STTfragment.kt]: If I make thisActivity lateinit var, it should be non null type. But activity is always null type and also cannot smart cast to activity.

**Comment by liveHarshit** [comment]
@iamareebjamal Updated. Sorry to use smart cast at each place instead of using some small logics.

File: app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/ChatFeedRecyclerAdapter.kt [hunk 17]
```diff
@@ -287,10 +287,10 @@ class ChatFeedRecyclerAdapter(private val currContext: Context, data: OrderedRea
      * @param text text to be copied
      */
     private fun setClipboard(text: String?) {
-        val clipboard = currContext.getSystemService(Context.CLIPBOARD_SERVICE) as android.content.ClipboardManager?
+        val clipboard = currContext.getSystemService(Context.CLIPBOARD_SERVICE)
         val clip = android.content.ClipData.newPlainText("Copied Text", text)
         if (clipboard != null) {
-            clipboard.primaryClip = clip
+            if (clipboard is android.content.ClipboardManager?) clipboard.primaryClip = clip
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/ChatFeedRecyclerAdapter.kt]: Revert to as it was

File: app/src/main/java/org/fossasia/susi/ai/device/DeviceActivity.kt [hunk 18]
```diff
@@ -28,7 +28,8 @@ class DeviceActivity : AppCompatActivity() {
         overridePendingTransition(R.anim.trans_left_in, R.anim.trans_left_out)
         setContentView(R.layout.activity_device)
 
-        mainWifi = application.getSystemService(Context.WIFI_SERVICE) as WifiManager
+        val wifiService = application.getSystemService(Context.WIFI_SERVICE)
+        if (wifiService is WifiManager) mainWifi = wifiService
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/device/DeviceActivity.kt]: As discusses, revert to original wherever getSystemService is used

**Comment by liveHarshit** [comment]
Updated

**Comment by liveHarshit** [comment]
@iamareebjamal Please review it, otherwise there will be conflicts with spotless changes.

File: app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/SelectionDialogListAdapter.kt (hunk 6 above)
Comment (iamareebjamal) [reply app/src/main/java/org/fossasia/susi/ai/chat/adapters/recycleradapters/SelectionDialogListAdapter.kt]: This was fine as it was. Just change the generic of findViewById from View to specific

File: app/src/main/java/org/fossasia/susi/ai/login/WelcomeActivity.kt [hunk 19]
```diff
@@ -64,8 +64,10 @@ class WelcomeActivity : AppCompatActivity() {
         ImageViewPager = findViewById<View>(R.id.pager) as ViewPager
         val tabLayout = findViewById<View>(R.id.tabDots) as TabLayout
         tabLayout.setupWithViewPager(ImageViewPager, true)
-        btnSkip = findViewById<View>(R.id.btn_skip) as Button
-        btnNext = findViewById<View>(R.id.btn_next) as Button
+        val viewSkip = findViewById<View>(R.id.btn_skip)
+        if (viewSkip is Button) btnSkip =  viewSkip
+        val viewNext = findViewById<View>(R.id.btn_next)
+        if (viewNext is Button) btnNext = viewNext
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/login/WelcomeActivity.kt]: Same as above

File: app/src/main/java/org/fossasia/susi/ai/skills/feedback/FeedbackActivity.kt [hunk 20]
```diff
@@ -17,7 +17,9 @@ class FeedbackActivity : AppCompatActivity() {
         super.onCreate(savedInstanceState)
         setContentView(R.layout.activity_feedback)
 
-        val feedbackResponse = intent.extras.get("feedbackResponse") as GetSkillFeedbackResponse
+        val extras = intent.extras.get("feedbackResponse")
+        var feedbackResponse: GetSkillFeedbackResponse? = null
+        if (extras is GetSkillFeedbackResponse) feedbackResponse = extras
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/feedback/FeedbackActivity.kt]: Revert to as it was

File: app/src/main/java/org/fossasia/susi/ai/skills/settings/ChatSettingsFragment.kt [hunk 21]
```diff
@@ -85,7 +86,8 @@ class ChatSettingsFragment : PreferenceFragmentCompat(), ISettingsView {
         speechOutput = preferenceManager.findPreference(getString(R.string.settings_speechPreference_key))
         speechAlways = preferenceManager.findPreference(getString(R.string.settings_speechAlways_key))
         displayEmail = preferenceManager.findPreference("display_email")
-        querylanguage = preferenceManager.findPreference(Constant.LANG_SELECT) as ListPreference
+        val preference = preferenceManager.findPreference(Constant.LANG_SELECT)
+        if (preference is ListPreference) querylanguage =  preference
```
Comment (iamareebjamal) [inline app/src/main/java/org/fossasia/susi/ai/skills/settings/ChatSettingsFragment.kt]: Revert

**Comment by liveHarshit** [comment]
Updated. The build is failing due to SDK 28 which was added in PR #1862. Build will pass after the PR #1862 got merged.

**Comment by liveHarshit** [comment]
Updated.

**Pull request merged**